scheduler/src/cook/mesos/util.clj

```diff
@@ -553,27 +553,35 @@
   (->> (repeatedly ch-size #(async/poll! ch))
        (remove nil?)))
 
-(defn xform-pipe
+(defn reduce-pipe
   "Reads elements from the `in-chan` channel and supplies elements to the `out-chan` channel.
    Maintains a state (initialized to `initial-state`) that is updated by applying the reducing function
    `in-xform` to the current state and the incoming element `(in-xform state element)`.
    `out-xform` accepts the current state (i.e. `(out-xform state)`) and produces `[reset-state out-value]`.
-   When the `out-chan` channel can receive an element, the `value` is put on the channel and the state is
-   set to the `reset-state` returned from the `out-xform`. Invocation of `out-xform` does not guarantee that
-   `out-value` will be consumed, hence it should not have side-effects. The optional parameter `on-consumed`
-   can be used to register notifications of when a value has been supplied to the `out-chan` channel.
+   When the `out-chan` channel can receive an element, a `out-data-chan` channel is put on the `out-chan`
+   channel. Then the `out-value` is put on this `out-data-chan` and the state is set to the `reset-state`.
+   Invocation of `out-xform` guarantees that `out-value` has been provided, hence it can be used
```

Comment:
I actually liked the on-consumed call back so you can separate your stateful actions from the actual processing.